File: 01.rb
data = File.read('01.txt')
# data = '91212129' # one test data
# data = '123123' # two test data

def one(str)
  arr = str.split('')
  total = 0
  arr.size.times do
    total += arr[0].to_i if arr[0] == arr[1]
    arr.rotate!
  end
  total
end

def two(str)
  arr = str.split('')
  total = 0
  mid = arr.size / 2
  arr.size.times do
    total += arr[0].to_i if arr[0] == arr[mid]
    arr.rotate!
  end
  total
end

# p one(data)
p two(data)

File: 01.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"strings"
	"strconv"
)

func one(arr []string) int {
	var i int
	arr_length := len(arr)
	arr = append(arr, arr[0])
	for v := 0; v < arr_length; v++ {
		if arr[v] == arr[v + 1] {
			temp_int, _ := strconv.Atoi(arr[v])
			i = i + temp_int
		}
	}
  return i
}

func two(arr []string) int {
	i := 0
	arr_len := len(arr)
	mid := (arr_len / 2)
	for v := 0; v <= mid; v++ {
		mid_index := (v + mid) % arr_len
		if arr[v] == arr[mid_index] {
			temp_int, _ := strconv.Atoi(arr[v])
			i = i + (temp_int * 2)
		}
	}
  return i
}

func main() {
  input, err := ioutil.ReadFile("01.txt")
  if err != nil {
		log.Fatal(err)
	}

	data := string(input)
	arr := strings.Split(data, "")

  one := one(arr)
	fmt.Println(one)
	two := two(arr)
	fmt.Println(two)
}